### PR TITLE
Fix internal methods

### LINKED ISSUE #591: Fix Internal Object Methods
**What is wrong with the current implementation?**
Currently are only set for ordinary objects, but not for array, string etc. Every object in the spec gets these internal methods an example is `[[DefineOwnProperty]]` for ordinary objects it calls `OrdinaryDefineOwnProperty` but for Array `[[DefineOwnProperty]]` is different when the property that is being defined is "length" it calls `ArraySetLength` this is what shrinks the array. This is what causes #557 bug.

We should avoid trait objects since it forces dynamic dispatch, and it also is not a viable solution.      
With trait objects we are trying to solve an inheritance problem, rust does not have a way to describe a base class where we have base class (with the the properties) and all the object (Array, etc) inherit from it. Also this would force dynamic dispatch this would slow down the code and it would not allow some compiler optimizations, we know the exact number of object types we don`t have do do this.

**My proposed changes:**
This is how I intended to do it, we have the internal method for example `[[set]]`, we also have the type definitions of what the object is `Array`, `Map`, etc. We have all we need to make this a static dispatch. So something like this:
```rust
impl Object {
	fn ordinary_set(&self /* , ... other parameters */) -> bool {
		// implementation
	}
	fn set(&mut self /* , ... other parameters */ ) -> bool {
		match self.data {
			ObjectData::Array => {
				// Arrays [[set]]
			}
			ObjectData::String(string) => {
				// Strings [[set]]
			}
			// This is for ordinay objects and for objects
			// that don't define a special implementation of [[set]]
			_ => self.ordinary_set(/* , ... args*/)
		}
	}
} 
```
This way of implementing allows the compiler to do some optimization that can't be done in dynamic dispatch. 

What do you think? any other better ideas? or improvements?

### PR DESCRIPTION
**Pull request by tofpie**
This Pull Request fixes #591 and #602.

It changes the following:

- All internal methods of `GcObject` take a `&Context` and return a `Result`
- Implementation of specific internal methods for `String`
- `Context` becomes externally immutable, with only interior mutability
- `LexicalEnvironment` becomes externally immutable, with only interior mutability

### CODE REVIEW

**Comment by Razican** [comment]
### Test262 conformance changes:
  | Test result | master count | PR count | difference |
  | :---------: | :----------: | :------: | :--------: |
  | Total | 78,497 | 78,497 | 0 |
  | Passed | 24,805 | 24,857 | **+52** |
  | Ignored | 15,587 | 15,587 | 0 |
  | Failed | 38,105 | 38,053 | **-52** |
  | Panics | 26 | 26 | 0 |
  | Conformance | 31.60 | 31.67 | **+0.07%** |

File: boa/src/builtins/json/mod.rs [hunk 1]
```diff
@@ -223,7 +223,8 @@ impl Json {
                 } else {
                     Some(
                         replacer
-                            .get_property(key)
+                            .get_property(key, context)
+                            .unwrap()
```
Comment (RageKnify) [inline boa/src/builtins/json/mod.rs:227]: I think using `expect` is preferable
```suggestion
                            .expect("'key' is one of the keys of 'replacer'")```

File: boa/src/object/internal_methods.rs [hunk 2]
```diff
@@ -51,31 +73,67 @@ impl GcObject {
...
+    ///  - [ECMAScript reference][spec]
+    ///
+    /// [spec]: https://tc39.es/ecma262/#sec-ordinarypreventextensions
+    #[inline]
+    pub fn ordinary_prevent_extensions(&mut self, _context: &Context) -> Result<bool> {
         self.borrow_mut().extensible = false;
-        true
+        Ok(true)
+    }
+    /// Delete a property.
```
Comment (RageKnify) [inline boa/src/object/internal_methods.rs:91]: Leave an empty line for clarity
```suggestion
    }
    
    /// Delete a property.
```

File: boa/src/object/internal_methods.rs [hunk 3]
```diff
@@ -401,32 +535,100 @@ impl GcObject {
...
+    pub fn ordinary_own_property_keys(&self) -> Vec<PropertyKey> {
         self.borrow().keys().collect()
     }
 
+    /// OwnPropertyKeys for ordinary objects
+    ///
+    /// More information:
+    ///   - [ECMAScript reference][spec]
+    ///
+    /// [spec]: https://tc39.es/ecma262/#sec-ordinaryownpropertykeys
```
Comment (RageKnify) [inline boa/src/object/internal_methods.rs:596]: The comment was copy-pasted, should mention `String Exotic Objects` instead.

**Review by RageKnify — CHANGES_REQUESTED** [review]
Minor mistakest to fix. Can you explain the reason for:
- `Context` becomes externally immutable, with only interior mutability

Sounds, reasonable just trying to understand the reasoning.

**Comment by tofpie** [comment]
All internal methods (as defined by the spec, that is `[[Get]]`, `[[Set]]` or `[[DefineOwnProperty]]` for instance) of a `GcObject` should have a parameter `&mut Context` as in some exotic objects implementation (such as `Proxy`) they can make calls to functions. If you add `&mut Context` to internal methods, then you must also add it to nearly all methods of `EnvironmentTrait` as the environment is sometimes itself an object (in `ObjectEnvironmentRecord` or `GlobalEnvironmentRecord`), but then each time the environment is modified (when declaring a variable or calling a function), then you have to borrow `Context` mutably twice, once to get the `environment` and another time when you call a method of `EnvironmentTrait`. And this is of course not allowed by the Rust compiler. And this is the same for `LexicalEnvironment`.

**Review by RageKnify — APPROVED** [review]

**Review by Razican — APPROVED** [review]

**Comment by HalidOdat** [comment]
I would like to try to fix the `context`/`LexicalEnvironment` mutable borrow problem, without introducing so much interior mutability.

Hopefully I'll be done with this by tomorrow and create a PR, now that I have a little more time to spend on the project :)

**Comment by Razican** [comment]
> I would like to try to fix the `context`/`LexicalEnvironment` mutable borrow problem, without introducing so much interior mutability.
> 
> Hopefully I'll be done with this by tomorrow and create a PR, now that I have a little more time to spend on the project :)

How is this going?

BTW, this needs a rebase with all the changes in master now.

**Comment by HalidOdat** [comment]
> > I would like to try to fix the `context`/`LexicalEnvironment` mutable borrow problem, without introducing so much interior mutability.
> > Hopefully I'll be done with this by tomorrow and create a PR, now that I have a little more time to spend on the project :)
> 
> How is this going?

#1131 is going to fix this problem, without any additional interior mutability :)

**Comment by Razican** [comment]
Now that #1131 has been merged, I think that this only needs a rebase to use the latest changes, @tofpie.

**Comment by HalidOdat** [comment]
This has too many merge conflicts to try rebase this, it would be better to start over, besides most of the internal methods have already been implemented. So closing this PR.

**Pull request closed without merging**